### DISCUSSION TITLE
Are chunk hashes not deterministic based on BUILD_ID?

### DISCUSSION THREAD

**Discussion by CaptainN** [vercel/next.js#65856] [Help]
### Summary

I have a site that we have been running with 2 horizontally scaled pods, which uses `generateBuildId` to return the git hash of the current commit for reliable scaling. I thought that meant the chunk hashes would be deterministically generated - that is, they'd be the same between pods (which are built at startup in the runtime environment, since static generation occurs at build time, and the necessary services are not available in the build environment).

This worked for a long time, but we just had a site meltdown because the chunk hash names didn't match, and when balancing occured, it was causing errors in loading. The solution was to scale back down to one instance for the time being. I did confirm by inspecting the build artifacts that .next/BUILD_ID matched between the pods, but the chunk names which were erroring did not.

Is BUILD_ID not intended to be used this way?

### Additional information

```js
I'm currently on next@14.1.4

Operating System:
  Platform: darwin (dev machine - we run node:20-alpine in production)
  Arch: arm64
  Version: Darwin Kernel Version 23.4.0: Fri Mar 15 00:10:42 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T6000
Binaries:
  Node: 20.11.1
  npm: 10.2.4 (this project uses npm)
  Yarn: 1.22.22
  pnpm: 8.15.4
Relevant Packages:
  next: 14.1.4
  eslint-config-next: 14.1.3
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.4.2
Next.js Config:
  output: N/A
```

**Comment by CaptainN** [comment]
I just did a quick test, and can confirm, the chunk names are not the same between builds. All I did is build once locally, delete the .next folder, then build again. The results are different chunk names:

```
# First Build:
❯ cat BUILD_ID
3cdb0cdf731951b8f03f4d8c16c600c5ead0d6eb

# in:
/.next/static/chunks/app/about/methodology/[slug]

❯ ls
page-fc743ef5cdb60c1d.js
```

```
# Second Build:
❯ cat BUILD_ID
3cdb0cdf731951b8f03f4d8c16c600c5ead0d6eb

# in:
/.next/static/chunks/app/about/methodology/[slug]

❯ ls
page-8acd4de2256ef601.js
```

Many of the build artifacts do have the same names - it seems to be these app chunks that don't get the same (deterministic) hashes.

**Reply by CaptainN** [reply]
Is this expected behavior, or is there a bug?

**Reply by CaptainN** [reply]
(Clearly a bug.)

**Comment by CaptainN** [comment]
bump

**Comment by Djay77** [comment]
Hey,

We have exactly the same problem here ! 
Multiples pods who build at startup and generate different chunks => some JS are in 404, then the page fall in error

What is the real utility of the Next.js config **generateBuildId** ... ?

**Reply by Djay77** [reply]
We have to override the webpack config in next.config file

```
webpack: (
        config,
        { buildId }
    ) => {
        config.output.filename = config.output.filename.replace('[chunkhash]', buildId);

        return config
    },
```

We have now the same name generated for our files BUT the content change

<img width="1082" alt="image" src="https://github.com/vercel/next.js/assets/3582550/416cde29-0537-46bb-b514-d958cd4745e5">

<img width="1079" alt="image" src="https://github.com/vercel/next.js/assets/3582550/71460461-38d9-452c-87d0-9b6c2145b133">

**Comment by CaptainN** [comment]
I tried to create an issue, but it requires a bunch of work since you have to create a reproduction repo. This is easy to understand, and should be in a unit test somewhere.

**Comment by austin-raey** [comment]
I'd like to concur on this issue,

The codebase I'm working with is quite complex so there is likely a lot of moving parts causing this discrepancy, however I would like to note:

I have a pretty simple layout file that is shared for a few pages, seen with this code:
```tsx
// layout-component.tsx
import Script from 'next/script';

import Footer from './footer';
import Header from './header';
import Main from './main';

export default function DefaultLayout({
  children,
}: Readonly<{
  children?: React.ReactNode;
}>) {
  return (
    <>
      {process.env.NODE_ENV === 'production' && <Script src="/analytics.js" />}
      <Header />
      <Main>{children}</Main>
      <Footer />
    </>
  );
}
```

With that code in that particular file, and a `generateBuildId` config present to always generate the same Build ID, I will consistently produce `page-e1ea5a6b17a0777a.js` in `.next/static/chunks/app/...`. I have replicated this running build many, many times.

However, simply removing the `<Script>` tag now makes all pages dependent on this have inconsistent filenames in `.next/static/chunks/app/...`, *all else held equal*:

```tsx
// layout-component.tsx, modified to remove <Script />.
import Footer from './footer';
import Header from './header';
import Main from './main';

export default function DefaultLayout({
  children,
}: Readonly<{
  children?: React.ReactNode;
}>) {
  return (
    <>
      <Header />
      <Main>{children}</Main>
      <Footer />
    </>
  );
}
```

Now, the `page.js` file will *cycle* between about 3 or 4 different filenames: `page-090c4b84bc1c1c00.js`, `page-c2924b123f71afcb.js`, `page-a711e6f388147d72.js`. It becomes semi-deterministic? 

Completely confounded as to why such a change would cause this issue, especially with a Build ID. Being that we are building to a couple different load balancers, this is a very irritating issue.

I had previously thought this non-deterministic issue was potentially caused by more complex `node_module` dependencies, but with this I have no idea.

**Comment by owenallenaz** [comment]
For what it's worth, what we do is docker build which runs the `next build` and then we re-use the image in two different Kube pods using `replicas: 2`. So then it's the **exact** same content, rather than needing to run build twice (which would just be a performance slog in CI anyways).

**Reply by austin-raey** [reply]
Yep, this is what we're looking at right now. Our operation had gotten by with 2 nodes, so we only needed 2 separate builds. This has generally worked o.k. in the past. 

The solution going forward is we're going to just do as you say -- build *once*, and in our case, export the resulting Docker Image to both nodes. This ensures content is always exactly the same regardless of Build IDs or whatever content hashes come up. I'm sure Kubernetes is geared for this exact use case but I'll have to review this in future 😵‍💫

Honestly, this also seems better practice anyways. :joy: I could imagine scaling to 5 or 10 nodes, and then having to run 5 or 10 separate builds. That doesn't seem optimal lol. 

Nonetheless, an interesting issue this is.

**Reply by CaptainN** [reply]
The problem with this is compile and generate are both done in the build context. In our deployment, we don't even have access to the services that would create the generate artifacts, since build and runtime are in two entirely different bits of infrastructure. But also, you aren't supposed to produce varying build artifacts - that's a docker no-no.

There is a kind of away around it, and it does actually address the bug described in this thread. You can do ["compile"](https://github.com/vercel/next.js/discussions/46544) in the build environment, which creates all the chunks, and therefor, can't vary the chunk names in the runtime. This is still a bug, and it should still be fixed, because who knows what else it's doing. Then you just have to do [generate](https://github.com/vercel/next.js/discussions/46544) in the runtime environment at startup. This also speeds up pod startup dramatically.

You still might end up with variable static generation artifacts, but you can work around that by using a [custom redis backing store or something](https://nextjs.org/docs/app/building-your-application/deploying#configuring-caching).

What would be even better is some straightforward way to just disable static generation at startup of all kinds, and let it fill the ISR cache at runtime. Maybe one day.

**Comment by JohnRPB** [comment]
This is a bit confounding to me as well. I'm using a consistent `generateBuildId()` output so my generated webpack files should all have the same names, but a few don't and cause key requests to fail with a 404 error. This means that caching pages across builds doesn't work; you have to delete the page entries in the cache on deployment. Not the behavior I was expecting from `generateBuildId()` :-/

**Comment by CaptainN** [comment]
I guess someone is going to have to make a reproduction repro, and file a ticket, even though there's more than enough information on this thread to see the problem...

@JohnRPB That used to work as expected. Now we get variable chunk names. There are workarounds available (described above), but none of that should be necessary.

**Comment by michael-meetkai** [comment]
hey @CaptainN @JohnRPB  have u found a solution ? having the same issue where the server chunks are different between my local env and CI/CD

**Reply by CaptainN** [reply]
My solution is to split build - compile in GHA, which locks the artifacts, and then do generation in startup. This works because the variable output happens in the compile stage. This relies on side effects being stable, which is not super amazing. It'd be nicer if the next.js team would fix the core issue.

**Comment by mhmmdglc** [comment]
is there any update ?

**Comment by samcarton** [comment]
Adding my 2c to this - This is a blocker especially given source maps are either all or nothing. If you want a public build without source maps, and a separate source-mapped build for external monitoring (sentry/datadog etc), the chunk names are going to be different. In this related discussion https://github.com/vercel/next.js/discussions/7579#discussioncomment-5092927 there are only half-solutions which involve stripping out sourcemaps from the deployed build which seems to cause 404s in some browsers.

**Comment by mhmmdglc** [comment]
I think Vercel is intentionally not solving this so that we use their platform :D

**Reply by CaptainN** [reply]
I don't think it's intentional, but at this point I would not recommend running next.js in production outside of vercel. It's clearly not well tested or maintained for that scenario.

([OpenNext](https://opennext.js.org/) exists, but it shouldn't have to.)

**Comment by junerd** [comment]
```
/** @type {import('next').NextConfig} */
const nextConfig = {
  output: 'export',
  productionBrowserSourceMaps: true,
  generateBuildId: () => 'static-build-id',
  webpack: (config) => {
    // Next.js output is not stable when there's no code change and this is a work around
    config.output.filename = config.output.filename.replace('[chunkhash]', '[contenthash]');
    return config;
  },
  devIndicators: false,
};
  ```
  
This worked for me I'm on `"next": "^15",`

**Reply by stevenlafl** [reply]
Interestingly, it behaves completely differently on different computers. Same files. `sam build` screws it up somehow.

So for CI builds:
```
const nextConfig: NextConfig = {
  output: 'standalone',
  eslint: {
    ignoreDuringBuilds: true,
  },
  typescript: {
    ignoreBuildErrors: true,
  },
  generateBuildId: () => {
    // Use GitHub Actions build number if available, otherwise fallback to 'build-id'
    return crypto.createHash('sha1').update(process.env.GITHUB_RUN_ID || 'build-id', 'utf8').digest('hex').substring(0, 7);
  },
  webpack: (config) => {

    // Next.js output is not stable when there's no code change and this is a work around
    config.output.filename = config.output.filename.replace('[chunkhash]', crypto.createHash('sha1').update(process.env.GITHUB_RUN_ID || 'build-id', 'utf8').digest('hex').substring(0, 7));
 
    return config;
  },
};
```

Invalidate your cache of course.

**Reply by junerd** [reply]
```
// this is madness, but it's working with next 15.2.4
function removeRandomness(el) {
  const prefix = `[\\"$\\",\\"$1\\",\\"`;
  const suffix = `\\"`;
  const prefixIndexes = [];
  let index = -1;
  do {
    if (typeof index === 'number') {
      prefixIndexes.push(index);
    }
    index = el.innerHTML.indexOf(prefix, index + 1);
  } while (index !== -1);
  prefixIndexes.shift();
  prefixIndexes.forEach((prefixIndex) => {
    const startIndex = prefixIndex + prefix.length;
    const endIndex = el.innerHTML.indexOf(suffix, startIndex);
    if (endIndex - startIndex < 2) {
      return;
    }
    const randomness = el.innerHTML.substring(startIndex, endIndex);
    const original = el.innerHTML.substring(prefixIndex, endIndex + suffix.length);
    el.innerHTML = el.innerHTML.replace(randomness, 'stable-id');

    const modified = el.innerHTML.substring(
      prefixIndex,
      el.innerHTML.indexOf(suffix, startIndex) + suffix.length
    );
    console.log(`👉 Replaced ${original} with ${modified} to stabilize build output`);
  });
}
```

I tracked it down to an inline script element, the code above works to stabilize the generated html for me now but yeah no idea why the the random string is added, couldn't tell by quickly looking at nextjs repo

**Reply by barroudjo** [reply]
Where do you put this code @junerd ?

On my end I mostly solved this issue with this workaround:
```javascript
const webpackConfig = (config, params) => {
    // config.optimization.realContentHash = true;
    if (config.name === 'client') {
      config.output.chunkFilename = `static/js/[name]-${params.buildId}.chunk.js`;
      config.output.filename = `static/js/[name]-${params.buildId}.js`;
      const index = /** @type {import('webpack').WebpackPluginInstance[]}  */ (
        config.plugins
      ).findIndex((plugin) => plugin instanceof NextMiniCssExtractPlugin);
      if (index !== -1) {
        const oldOptions = config.plugins[index].options;
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        config.plugins[index] = new NextMiniCssExtractPlugin({
          ...oldOptions,
          filename: `static/css/[name]-${params.buildId}.css`,
          chunkFilename: `static/css/[name]-${params.buildId}.css`,
        });
        console.log(
          '!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Updated MiniCssExtractPlugin options',
        );
      }
    }
    return config;
  }
```

**Reply by junerd** [reply]
I think that might be a different issue since the chunk (js or css) files themselves are stable for me but I did have to alter the next.js.config to
```
/** @type {import('next').NextConfig} */
const nextConfig = {
  output: 'export',
  productionBrowserSourceMaps: true,
  generateBuildId: () => 'static-build-id',
  webpack: (config) => {
    // Next.js output is not stable when there's no code change and this is a work around
    config.output.filename = config.output.filename.replace('[chunkhash]', '[contenthash]');
    return config;
  },
  devIndicators: false,
};
```

The `removeRandomness` function is in a custom post build script which parses the generated index.html and modifies that script element

**Reply by barroudjo** [reply]
Thanks ! What a mess, you'd think having a deterministic build would be a priority for the Next team...

**Reply by cmdcolin** [reply]
for what it's worth, I managed to get @junerd's code to output stable ID like this: months later edit: removed. It didn't work. I will leave the details in a `<details>` below but know that it did not work as far as i could tell, there was still randomness in the hashes

<details>
 I asked AI to generate a script to run the replaceRandomness function with jsdom

```#!/usr/bin/env node

// replaceRandomness function credit @junerd
// https://github.com/vercel/next.js/discussions/65856#discussioncomment-12667717
function removeRandomness(el) {
  const prefix = `[\\"$\\",\\"$1\\",\\"`
  const suffix = `\\"`
  const prefixIndexes = []
  let index = -1
  do {
    if (typeof index === 'number') {
      prefixIndexes.push(index)
    }
    index = el.innerHTML.indexOf(prefix, index + 1)
  } while (index !== -1)
  prefixIndexes.shift()
  prefixIndexes.forEach(prefixIndex => {
    const startIndex = prefixIndex + prefix.length
    const endIndex = el.innerHTML.indexOf(suffix, startIndex)
    if (endIndex - startIndex < 2) {
      return
    }
    const randomness = el.innerHTML.substring(startIndex, endIndex)
    const original = el.innerHTML.substring(
      prefixIndex,
      endIndex + suffix.length,
    )
    el.innerHTML = el.innerHTML.replace(randomness, 'stable-id')

    const modified = el.innerHTML.substring(
      prefixIndex,
      el.innerHTML.indexOf(suffix, startIndex) + suffix.length,
    )
    // console.log(
    //   `👉 Replaced ${original} with ${modified} to stabilize build output`,
    // )
  })
}

/**
 * Example script to demonstrate how to use the stabilizeHtml.js file
 * This script processes an HTML file and removes randomness from it
 *
 * Usage: node processHtmlExample.js <input-html-file> [output-html-file]
 */

// Get command line arguments
const args = process.argv.slice(2)

if (args.length < 1) {
  console.error(
    'Usage: node processHtmlExample.js <input-html-file> [output-html-file]',
  )
  process.exit(1)
}

const inputFilePath = args[0]
const outputFilePath =
  args.length > 1 ? args[1] : inputFilePath.replace('.html', '.stable.html')

// Process the HTML file
// console.log(`Processing HTML file: ${inputFilePath}`)
// console.log(`Output will be saved to: ${outputFilePath}`)

// Import required modules
const fs = require('fs')
const path = require('path')
const { JSDOM } = require('jsdom')

/**
 * Processes an HTML file to remove randomness
 * @param {string} inputFilePath - Path to the input HTML file
 * @param {string} outputFilePath - Path to save the processed HTML file (optional)
 * @returns {Promise<string>} - The processed HTML content
 */
async function processHtmlFile(inputFilePath, outputFilePath) {
  try {
    // Read the HTML file
    const htmlContent = fs.readFileSync(inputFilePath, 'utf8')

    // Parse the HTML using JSDOM
    const dom = new JSDOM(htmlContent)
    const document = dom.window.document

    // Find all elements in the document
    const allElements = document.querySelectorAll('*')

    // Apply removeRandomness to each element
    allElements.forEach(el => {
      // Create a wrapper with innerHTML property for compatibility with the function
      const wrapper = {
        innerHTML: el.outerHTML,
      }

      // Apply the removeRandomness function
      removeRandomness(wrapper)

      // If the innerHTML was changed, update the element
      if (wrapper.innerHTML !== el.outerHTML) {
        // Replace the element with the modified version
        const tempDiv = document.createElement('div')
        tempDiv.innerHTML = wrapper.innerHTML
        el.parentNode.replaceChild(tempDiv.firstChild, el)
      }
    })

    // Get the processed HTML
    const processedHtml = dom.serialize()

    // Save to output file if specified
    if (outputFilePath) {
      fs.writeFileSync(outputFilePath, processedHtml, 'utf8')
      console.log(`Processed HTML saved to ${outputFilePath}`)
    }

    return processedHtml
  } catch (error) {
    console.error('Error processing HTML file:', error)
    throw error
  }
}

processHtmlFile(inputFilePath, outputFilePath)
  .then(() => {
    console.log('HTML processing completed successfully')
  })
  .catch(error => {
    console.error('Failed to process HTML:', error)
    process.exit(1)
  })

```

then I run this over all files in the "out" directory (output of nextjs export:static build) with fd-find+gnu parallel    

```
 "postbuild": "fd html out| parallel \"node createReproducibleHtml.js {} {}\"",
```


now, when I rebuild the site, I can use rclone to upload to AWS S3 and it only uploads the couple files that were updated.... (note: I use `rclone sync` because it is able to do checksums for file upload,  `aws s3 sync` command only does file-last-modified which doesn't work since of course all file times are modified after build! rclone command with -c which is checksum: `rclone sync -c -v out mysite:bucketname/folder/`)
</details>

**Comment by dmitrc** [comment]
@leerob, the issue described in this thread is a serious blocker to scalable self-hosting.
Given the team's focus on deployment options and level playing field, is this something that can be looked at?

**Reply by junerd** [reply]
Combination of https://github.com/vercel/next.js/discussions/65856#discussioncomment-12387797 and https://github.com/vercel/next.js/discussions/65856#discussioncomment-12667717 made it stable for me. The second one is pretty hacky though I have to say

**Reply by CaptainN** [reply]
The recent RFP was about stablizing the alternative build internals - the build apis that Vercel uses, for it's deployment strategy. The idea there is that OpenNext partners can use that, instead of relying on next-server.

The issue is that OpenNext wouldn't use the undocumented internal APIs that Vercel uses, because they were undocumented, and potentially unstable. But now they can.

The key takeaway is that FEWER key players will be using the standalone next-server, not more. I would not count on patches to next-server, not even for key problems like the one described here, or the various memory and even CPU cycle leaks. They are unlikely to get patched.

**Reply by dmitrc** [reply]
I am not sure how one relates to other. These are still the build outputs, whether they are declared in the manifest for the new adapters API, or not. It will still not solve the non-determinism of these builds when running multiple times for different stages etc, and result in the same category of issues with 404s from CDN and more. Not to mention we don't have a crystal ball re: the end usage of deployment solutions, and manual Docker/Kubernetes will always remain desirable for a non-negligible subset of users.

We shouldn't have to resort to these hacks that @junerd in describing, when `generateBuildId`'s whole purpose is to handle cases like these.

**Reply by CaptainN** [reply]
To be clear - this is a bug. It's not a feature. The way to verify it is easy (use create next app to create a new app, then build it multiple times and note how the hash names change with each build, perhaps with a small subset of possible outputs.) It's just a question of whether the bug will be prioritized and fixed.

**Comment by sidoruk-sv** [comment]
Sorry, but wow!

Year 2025

The popular js/react "framework" still can not produce deterministic builds, so only hacks and post-build scripting save 😿

**Reply by GabenGar** [reply]
Works on vercel.

**Reply by sidoruk-sv** [reply]
@GabenGar does not work on my computer

**Comment by CNsixgod** [comment]
Does this happen in turbo build Mode?

**Reply by sidoruk-sv** [reply]
~no, only Webpack variant is affected (so it's hash generation is based not on a content of the files but also includes timestamp)~

rechecked myself and I see the amount of changed files is less for turbopack, but still because the folder name (which contains files like _buildManifest.js and _ssgManifest.js) are changing every build, we have a lot of "affected files". Even though _buildManifest.js and _ssgManifest.js are identical in both folders

**Comment by sidoruk-sv** [comment]
My concern here is not about reproducible builds across different machines (though that would be nice to have in 2025).

The real issue for me is that we deploy new releases to devices over very slow and unreliable internet connections. To save bandwidth, we try to upload only changed files between builds. But since the Next.js build output is not deterministic, file hashes often change even when the content hasn’t, which forces us to re-upload almost the entire `out` folder (we have `output: 'export'`) unnecessarily.

**Reply by cmdcolin** [reply]
@sidoruk-sv I had same issue with static site builds reuploading tons of files. I switched to astro to solve. not ideal, and I like the next.js ethos, but it was at the point where every build reuploaded 1.5 gigabytes to s3

**Comment by long2ice** [comment]
is there any update ?

**Comment by XampleV** [comment]
Any update on this?

**Comment by sefin23** [comment]
This is a confirmed bug: Next.js chunk hashes are **not deterministic** between builds, even when `generateBuildId` returns the same value. Here's everything the thread has uncovered, plus the current best workarounds:

## Root Cause

The non-determinism comes from **two sources**:

1. **Webpack's `[chunkhash]`** — uses a hash algorithm that includes non-content factors (like module ordering, timestamps, or internal IDs), not purely file content. So identical code can produce different hashes across builds.
2. **Inline script randomness** — Next.js injects an inline `<script>` tag with a randomly-generated string into the HTML output (tracked down by @junerd). This causes downstream hashes to vary even if your code is unchanged.

This directly breaks:
- **Horizontal scaling** (multiple pods built separately — each gets different chunk names, causing 404s when load-balanced)
- **Incremental deployments** (diffing build outputs for delta uploads)
- **Multi-stage builds** (build for Sentry source maps separately from production deployment)
- **Cross-build caching**

## Workarounds (from this thread)

### ✅ 1. Build once, share the artifact (recommended for horizontal scaling)

The simplest and most reliable fix: **build once in CI, then distribute the same artifact to all pods**.

```dockerfile
# Build once
docker build -t myapp:latest .

# Then run multiple replicas using the exact same image
# Kubernetes: set replicas: 2+ pointing to same image
```

This guarantees identical chunk names across all pods since they all use the same build output. @owenallenaz confirmed this works well.

### ✅ 2. Split compile vs. generate (for runtime-only static generation)

As described by @CaptainN — if your setup requires static generation to happen at pod startup (because build-time services aren't available):

1. Run `next build` (compile stage only) in CI → this locks all chunk names
2. At pod startup, only run the generate phase

See the [compile/generate split discussion](https://github.com/vercel/next.js/discussions/46544). This dramatically speeds up pod startup too.

### ✅ 3. Replace `[chunkhash]` with `[contenthash]` in webpack config

@junerd found this makes chunk filenames stable (content-based, not build-order-based):

```js
// next.config.js
module.exports = {
  generateBuildId: () => 'your-stable-build-id', // e.g. git SHA
  webpack: (config) => {
    config.output.filename = config.output.filename.replace('[chunkhash]', '[contenthash]');
    return config;
  },
};
```

`[contenthash]` hashes only the file's content, making it deterministic across builds when code hasn't changed.

### ✅ 4. Override chunk and CSS filenames with buildId (for full determinism)

@barroudjo's more complete approach — forces all client chunk filenames to be based on the `buildId`:

```js
// next.config.js
const webpackConfig = (config, { buildId }) => {
  if (config.name === 'client') {
    config.output.chunkFilename = `static/js/[name]-${buildId}.chunk.js`;
    config.output.filename = `static/js/[name]-${buildId}.js`;
    // Also update MiniCssExtractPlugin if you use CSS
  }
  return config;
};
```

Note: This makes filenames stable but the file **content** may still vary if there's inline randomness (see point 2 of root causes above).

### ⚠️ 5. Turbopack (`next build --turbo`)

@sidoruk-sv checked: Turbopack produces fewer changed files per build, but `_buildManifest.js` and `_ssgManifest.js` still change between builds even with identical content, so it doesn't fully solve the problem either.

## Current Status (as of March 2026)

- This bug **still exists** in Next.js 15+ and even 16
- No official fix or acknowledgment from the Next.js team
- The `generateBuildId` option's documentation implies deterministic builds but doesn't deliver them
- @CaptainN's assessment: this is unlikely to be prioritized since Vercel's own infrastructure doesn't suffer from it (they deploy via their platform which sidesteps this entirely)

## Recommended approach by use case

| Use case | Best fix |
|---|---|
| Horizontal scaling (K8s/Docker) | Build once, share same image across replicas |
| Runtime static generation needed | Compile in CI, generate at pod startup |
| Delta deployments / CDN caching | Use `[contenthash]` webpack override |
| Static export (`output: 'export'`) | `[contenthash]` override + post-build script for inline randomness |

For most production setups, **building once and reusing the artifact** is the cleanest solution and the standard Docker/Kubernetes pattern anyway.

**Reply by CaptainN** [reply]
A quick addition - since writing all this, next.js team added a `generate-env` stage, which you can use in place of `generate` - it only does the ENV var inlining step, and doesn't do ANY static generation, which I personally find ideal. Since it delays all generation to the ISR cache, it takes less than one second to start up the pod - vastly improved ver either compile + generate or even generate alone.
